src/navigation/data_collection.py

```diff
@@ -0,0 +1,179 @@
+import time
+import numpy as np
+import pandas as pd
+import rospy
+from sensor_msgs.msg import JointState
+from std_msgs.msg import Bool
+import datetime
+from util.SE3 import SE3
+from util.SO3 import SO3
+from threading import Lock
+from pathlib import Path
+import os
+
+
+def make_filename(csv_data):
+    # makes filename based on current time. "output_mmddyyyy_hr-min-sec.csv"
+    now = datetime.datetime.now()
+    day = now.strftime("%m%d%Y")
+    hour = now.strftime("%H-%M-%S")
+    time_stamp = day + "_" + hour
+    home = str(Path.home())
+    folder = home + "/catkin_ws/src/mrover/failure_data"
+    if not os.path.exists(folder):
+        os.makedirs(folder)
+    file = folder + "/output_" + time_stamp + ".csv"
+    rospy.logerr(f"Created {file} in data_collection.py")
+    return file
+
+
+class Data:
```

Comment:
This is exactly where `@dataclass` design pattern is really good. The object is a plain old data container and it should be immutable (frozen) and have one true constructor. If you need fancy constructors i'd build it as a `@classmethod`. Check out `SearchTrajectory` [here](https://github.com/umrover/mrover-ros/blob/dcabe6fcc82e1f714a28f91b80305a0d39e52c00/src/navigation/search.py#L17) for an example

Reply:
On second thought, do you actually want this to be a pandas `DataFrame`? I'm not sure if this is functionally any different and a `DataFrame` gives you the ability to index columns by label and extract numpy arrays for processing.